src/main/host/descriptor/socket/netlink.rs

```diff
@@ -0,0 +1,1215 @@
+use std::io::{Cursor, ErrorKind, Read, Write};
+use std::net::Ipv4Addr;
+use std::sync::{Arc, Weak};
+
+use atomic_refcell::AtomicRefCell;
+use linux_api::errno::Errno;
+use linux_api::ioctls::IoctlRequest;
+use neli::consts::nl::{NlmF, NlmFFlags, Nlmsg};
+use neli::consts::rtnl::{
+    Arphrd, Ifa, IfaF, IfaFFlags, Iff, IffFlags, Ifla, RtAddrFamily, RtScope, Rtm,
+};
+use neli::nl::{NlPayload, Nlmsghdr};
+use neli::rtnl::{Ifaddrmsg, Ifinfomsg, Rtattr};
+use neli::types::{Buffer, RtBuffer};
+use neli::{FromBytes, ToBytes};
+use nix::sys::socket::{MsgFlags, NetlinkAddr, Shutdown};
```

Comment:
I think we are also trying to stay away from `nix` usage like this, even though we had been using this a lot in other parts of our code. The [coding section in the docs](https://shadow.github.io/docs/guide/coding.html#platform-libc-and-linux-crates) explains a bit more.

I might be wrong, but I _think_ the use case here is to handle args ultimately coming from a syscall hander. In these cases I think we want to add these to our `linux-api` crate too. We only started building up the `linux-api` crate recently so it doesn't have everything we need yet. But[the time module](https://github.com/shadow/shadow/blob/main/src/lib/linux-api/src/time.rs) shows an example of what goes there.

@sporksmith can you confirm?

Reply:
Confirmed

Reply:
I saw that `MsgFlags` and `Shutdown` are also used in `socket/inet/tcp.rs` and `NetlinkAddr` is analogous to `SockaddrIn` in `socket/inet/tcp.rs`.

If I shouldn't use `nix`, I don't have an example to follow and I guess it's a new thing for Shadow as well.

They are not used only in tcp.rs but also udp.rs and unix.rs as well.

In addition, if I want to remove `Shutdown`, the socket interface has to be changed as well. See https://github.com/shadow/shadow/blob/main/src/main/host/descriptor/socket/mod.rs#L315

Reply:
Yeah while we are trying to get rid of nix, it's probably best to use nix for at least `Shutdown` which is part of the "socket interface", and you can't change it here without also needing to change it in all other socket types as well.

For `NetlinkAddr`, it would be nice to move away from nix as well (for example see `SockaddrUnix` in `main/utility/sockaddr.rs`; this was written before linux-api existed). But re-implementing `NetlinkAddr` in linux-api might be a bit of work, so I don't think it's worth holding up this PR over.

For `MsgFlags`, it would probably be straightforward to create this as a bitflags type in `linux-api`, but I don't have much opinion on this. If you want to do this, the steps are probably (1) Add `#include <linux/socket.h>` to `lib/linux-api/bindings-wrapper.h`, (2) run `lib/linux-api/gen-kernel-bindings.sh` to generate new bindings for the kernel api, (3) add a `lib/linux-api/src/socket.rs` file, and (4) create a new `MsgFlags` bitflags following an example like `OFlag` from `lib/linux-api/src/fcntl.rs`. You could find the `MsgFlags` flags by searching `lib/linux-api/src/bindings.rs` for constants that begin with `MSG_`.

Reply:
> For `MsgFlags`, it would probably be straightforward to create this as a bitflags type in `linux-api`

Actually Linux doesn't expose the `MSG_` flags. I don't think we have decided on a way to expose things in `linux-api` that the Linux headers don't expose. So probably best to keep using `nix::MsgFlags` unless someone else here gives better advice about how to handle this.

Reply:
I guess the conclusion is that we should use `nix` for all these three things.

Reply:
Yeah, continuing to use `nix` for these probably makes sense for this PR.

> Actually Linux doesn't expose the MSG_ flags. I don't think we have decided on a way to expose things in linux-api that the Linux headers don't expose.

For when we do want to tackle this - the way `linux-raw-sys` handles this is by having some manually created header files with the missing definitions. https://github.com/sunfishcode/linux-raw-sys/blob/main/gen/modules/net.h

We've done the same for some things, currently just in the single "wrapper" header file: https://github.com/shadow/shadow/blob/main/src/lib/linux-api/bindings-wrapper.h#L28

Eventually we may want to have `linux-api` just be high-level wrappers around the `linux-raw-sys` bindings instead of generating our own, but we currently still need our custom symbol name mangling to support some C code.